doc/developer/design/20210714_reclocking.md

```diff
@@ -0,0 +1,98 @@
+For example,
+    Kafka topics contain multiple "parts", each of which advances through "offsets".
+    Postgres sources come from a log where entries have a "log sequence number" (LSN).
+    File sources have a line number.
+    Materialize CDC sources have explicit timestamps.
+
+The source-specific "timestamps" of forward progress are useful, but they are not directly comparable.
+This document describes one framework for converting between the timestamps including in to and potentially out of Materialize's system time.
+
+## Goals
+
+The primary goal of this document is to describe a formal way to translate collections that evolve with one gauge into collections that evolve with another gauge.
+That formalism should have the properties that:
+1. There is an explicit representation of the translation (rather than a behavioral description).
+2. The translation should be sidecar metadata, rather than a rewriting of the source data.
+3. The combination of durable source data and durable metadata should result in a durable reclocked stream.
+4. The sidecar metadata should be compactable whenever the reclocked stream is compactable.
+
+The formalism should support the goals of
+A. Permitting the use of streams from one timeline in another timeline.
+B. Describing durable state that can make sources like Kafka, Postgres, and files NONVOLATILE in the strong sense that their contents can be replayed exactly.
+C. A BYO format for users who want to provide explicit timestamp information for sources that are presented without timestamps (or with timestamps from another time domain).
+D. A way to describe in one time domain progress made through another time domain (e.g. reporting Kafka read progress in MZ system timestamps).
+
+## Non-Goals
+
+This document has no opinions on the physical representation of the reclocking data.
+This document has no opinions on the implementation of the reclocking actions.
+This document has no opinions on the policies of which streams should be reclocked into which domains.
+
+## Description
+
+Streams of data may come with timestamps attached to each record, and a mechanism for determining when timestamps are complete.
+This information describes "progress" through a stream, and allows consumers of the stream to rely on the presence of certain contents.
+Multiple events may share the same timestamp, indicating that they should occur at the same moment.
+Messages may not be in timestamp order, but the stream should not claim that a time is complete until all messages with lesser timestamps have been received.
+
+### `remap`
+
+We are going to use a `remap` collection, which is a map from a target time domain `Into` to an antichain of timestamps from the source time domain `From`.
+The easiest way for me to represent this is as a differential dataflow collection,
+```rust
+// G's timestamp should be `Into`.
+remap: Collection<G, Antichain<From>>
```

Comment:
Why does this need to be `Antichain<From>`, rather than just `From` ?  Since in a collection there can be multiple entries at a given timestamp anyway.

Reply:
You are right. I think I wrote it first as a `HashMap`, and then backported the ideas to a collection.